Optimize StyledElement::rebuildPresentationalHintStyle()

cdumez opened the pull request:
#### 8249ec2a7b199f145256e194ad2fc97c61dbe404
<pre>
Optimize StyledElement::rebuildPresentationalHintStyle()
<a href="https://bugs.webkit.org/show_bug.cgi?id=258316">https://bugs.webkit.org/show_bug.cgi?id=258316</a>
rdar://111406451

Reviewed by Darin Adler.

Optimize StyledElement::rebuildPresentationalHintStyle() by leveraging
the NodeName enumeration to avoid a large HashMap lookup.

* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::rebuildPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
(WebCore::StyledElement::additionalPresentationalHintStyle const):
(WebCore::StyledElement::collectExtraStyleForPresentationalHints): Deleted.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::cssPropertyIdForSVGAttributeName):
(WebCore::createAttributeNameToCSSPropertyIDMap): Deleted.

Canonical link: <a href="https://commits.webkit.org/267482@main">https://commits.webkit.org/267482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58a49db053dfd69fff4cc040d1fd009a10244875

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17383 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18396 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15582 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17078 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17910 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17211 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19179 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15065 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21842 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19532 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15827 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13441 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15023 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4000 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19393 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/svg/SVGElement.cpp:
```diff
@@ -925,8 +833,152 @@ CSSPropertyID SVGElement::cssPropertyIdForSVGAttributeName(const QualifiedName&
 
-    static NeverDestroyed properties = createAttributeNameToCSSPropertyIDMap();
-    return properties.get().get(attrName.localName());
+    switch (attrName.nodeName()) {
```
Why not use a `SortedArrayMap`? It was designed for this, and could possibly be more efficient than a switch statement. Maybe you tried it and it was less efficient? Or maybe it's hard to keep the list in numeric order?

cdumez replied:
In the past whenever I tried to replace something with a SortedArrayMap in a hot code path, it ended up being slightly slower. I haven't tried this particular case though. Do we really think SortedArrayMap (with linear search) could be faster than a switch statement?

I'll find some time to try it out but I am not super hopeful.

darinadler replied:
Maybe don't waste your time.

One reason it has a **chance** to be faster is that it switches to binary search when the array is long enough, but I supposed we would **hope** that the compiled code for case statements is always so efficient it would be hard to beat. In theory a good version of `SortedArrayMap` ought to be able to detect when it's all constant at compile time and generate code with identical performance to a case statement, but perhaps that is fantasy.

cdumez committed:
Optimize StyledElement::rebuildPresentationalHintStyle()
https://bugs.webkit.org/show_bug.cgi?id=258316
rdar://111406451

Reviewed by Darin Adler.

Optimize StyledElement::rebuildPresentationalHintStyle() by leveraging
the NodeName enumeration to avoid a large HashMap lookup.

* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::rebuildPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
(WebCore::StyledElement::additionalPresentationalHintStyle const):
(WebCore::StyledElement::collectExtraStyleForPresentationalHints): Deleted.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::cssPropertyIdForSVGAttributeName):
(WebCore::createAttributeNameToCSSPropertyIDMap): Deleted.

Canonical link: https://commits.webkit.org/267482@main
